### PR TITLE
Add fuel consumption test

### PR DESCRIPTION
**Pull request by michael-weigelt**
Including the test from https://raw.githubusercontent.com/bytecodealliance/wasmtime/main/examples/fuel.c

### CODE REVIEW

File: lib/Wasmtime.hs [hunk 1]
```diff
@@ -445,6 +448,27 @@ withContext :: Context s -> (Ptr C'wasmtime_context_t -> IO a) -> IO a
...
+      then pure Nothing
+      else Just <$> peek amount_ptr
+
+consumeFuel :: Context s -> Word64 -> IO (Either WasmtimeError Word64)
+consumeFuel ctx amount = withContext ctx $ \ctx_ptr ->
+  alloca $ \remaining_ptr -> do
+    error_ptr <- c'wasmtime_context_consume_fuel ctx_ptr amount remaining_ptr
+    checkWasmtimeError error_ptr
+    Right <$> peek remaining_ptr
+
```
Comment (michael-weigelt) [inline lib/Wasmtime.hs:471]: should they be in ST instead of IO?

File: lib/Wasmtime.hs [hunk 2]
```diff
@@ -459,6 +462,27 @@ withContext :: Context s -> (Ptr C'wasmtime_context_t -> IO a) -> IO a
 withContext ctx f = withStore (storeContextStore ctx) $ \_store_ptr ->
   f $ storeContextPtr ctx
 
+addFuel :: Context s -> Word64 -> IO (Either WasmtimeError ())
+addFuel ctx amount = withContext ctx $ \ctx_ptr -> do
+  error_ptr <- c'wasmtime_context_add_fuel ctx_ptr amount
+  checkWasmtimeError error_ptr
+  pure $ Right ()
```
Comment (basvandijk) [inline lib/Wasmtime.hs:469]:
```suggestion
addFuel ctx amount = withContext ctx $ \ctx_ptr -> try $ do
  error_ptr <- c'wasmtime_context_add_fuel ctx_ptr amount
  checkWasmtimeError error_ptr
```

File: lib/Wasmtime.hs (hunk 1 above)
Comment (basvandijk) [reply lib/Wasmtime.hs:485]: Yes, let's do it.

File: lib/Wasmtime.hs [hunk 3]
```diff
@@ -459,6 +462,27 @@ withContext :: Context s -> (Ptr C'wasmtime_context_t -> IO a) -> IO a
...
+    if not res
+      then pure Nothing
+      else Just <$> peek amount_ptr
+
+consumeFuel :: Context s -> Word64 -> IO (Either WasmtimeError Word64)
+consumeFuel ctx amount = withContext ctx $ \ctx_ptr ->
+  alloca $ \remaining_ptr -> do
+    error_ptr <- c'wasmtime_context_consume_fuel ctx_ptr amount remaining_ptr
+    checkWasmtimeError error_ptr
+    Right <$> peek remaining_ptr
```
Comment (basvandijk) [inline lib/Wasmtime.hs:484]:
```suggestion
consumeFuel ctx amount = withContext ctx $ \ctx_ptr ->
  alloca $ \remaining_ptr -> try $ do
    error_ptr <- c'wasmtime_context_consume_fuel ctx_ptr amount remaining_ptr
    checkWasmtimeError error_ptr
    peek remaining_ptr
```

File: test/fuel.hs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  trapCode trap @?= Just TRAP_CODE_OUT_OF_FUEL
+
+  putStrLn "Shutting down..."
+  putStrLn "Done."
+
+handleWasmtimeError :: Either WasmtimeError a -> IO a
+handleWasmtimeError = either throwIO pure
+
+handleException :: Exception e => Either e r -> IO r
+handleException = either throwIO pure
```
Comment (basvandijk) [inline test/fuel.hs:59]: Could we just have a single:
```suggestion
handleException :: Exception e => Either e r -> IO r
handleException = either throwIO pure
```

**Review by basvandijk — APPROVED** [review]

**Pull request merged**